### ISSUE TITLE
LiveView support 

### ISSUE THREAD

**Issue by mayel** [elixir-waffle/waffle#71]
With the new upload functionality bundled in Phoenix LiveView, I wonder if anyone has successfully used Waffle with it, and if there are any gotchas or changes needed in the lib?

**Comment by achedeuzot** [comment]
I've currently used Waffle (without Waffle Ecto) with Phoenix LiveView.

I followed the phoenix blog & screencast https://www.phoenixframework.org/blog/phoenix-live-view-upload-deep-dive with a few changes:
- I've used the Mime type validation code which is in PR #69 but I now have a two step process (one for validating, one for consuming the uploads after the DB update)
- I couldn't see how it would work with `waffle_ecto` so I've implemented my own logic on top of a simple `string` ecto field.

The rest was quite straightforward to implement :)

**Comment by achempion** [comment]
@achedeuzot Could you please describe what issues did you have with `waffle_ecto` library? It provides couple of convenient functions like `cast_attachements` for the changeset as well as cache management and url generation which you still need for the LiveView.

**Comment by achedeuzot** [comment]
@achempion Indeed, `cast_attachements` seemed perfect for what I wanted but I couldn't see how to make it work with the `consume_uploaded_entries` function from LiveView if I followed the sample video from the Phoenix blog. I moved the `consume_uploaded_entries` before the call to the changeset to try and give the file path into `cast_attachements` but I couldn't make it work.

If I look at the code of `waffle_ecto`, I can see `cast_attachements`  accepts a `%Plug.Upload{}` which I don't have (LiveViews provides a `%LiveView.UploadEntry{}`), a `%{filename: filename, binary: binary}` map which I don't have either (and I don't want to read the file into the `:binary` key for memory consumption reasons) or a simple file path which looks like what I want but fails with `:invalid_path` even if I set `accept_paths: true`, the file path given by `consume_uploaded_entries` gives something like `/var/folders/e4/rnd23gsflk23_ynglbkeshdq9_m02340gq/T//plug-1704/live_view_upload-094958382-29457213941-4`.

Right now my steps are the following:
1 - validate schema params through a standard `changeset` function
2 - validate uploaded files through the `MyApp.AvatarUploader.validate/1` function and add errors into the changeset if needed
3 - save the record to database if everything is OK
4 - save the file to storage if everything is OK. If not, I'm rolling back the field in the DB.

I also couldn't see where I could call a custom `validate` function on the passed uploaded file using `cast_attachements`.

Finally, the way the `MyApp.AvatarUploader` module needs to contain `use Waffle.Ecto.Definition` and the schema must contain `field: :avatar, MyApp.AvatarUploader.Type` created a dependency schema which doesn't work well with my current application setup: the `MyApp.AvatarUploader` lives in an app that depends on the Ecto schema app but not the other way around. So it's creating a circular dependency between apps I'd rather not create.

**Comment by montebrown** [comment]
Hi @achedeuzot,

You can use waffle_ecto - I was able to get this working without changing my schema. You can construct a %Plug.Upload{} and pass it in to the changeset.  Something like:

```
def handle_event("save", _params, socket) do
  consume_uploaded_entries(socket, :avatar, fn meta, entry ->
    {:ok, user} =
      Accounts.update_user(%{
        "file" => %Plug.Upload{
          content_type: entry.client_type,
          filename: entry.client_name,
          path: meta.path
        },
      })

    Snap.Uploaders.Upload.url({upload.file, upload}, :original)
  end)

  {:noreply, socket |> update(:uploaded_files, &(&1 ++ uploaded_files)) |> fetch_user()}
end
```

**Comment by TomEversdijk** [comment]
Thanks @montebrown for your solution! 

For everybody who tried to get this working it is important to use the `cast_attachements` method within the `consume_uploaded_entries`. The `consume_uploaded_entries` will automatically delete the file-path when it is finished so otherwise and `{:error, :invalid_file_path}` error will be returned.

**Comment by Papipo** [comment]
@montebrown solution doesn't play well with multiple upload entries, does it?
If I want to leverage waffle_ecto and have access to the id of the parent schema to ensure that all images associated with it are stored in a folder by its id, I would need to update the record as many times as images were submitted?

At this point, given that I am using UUIDs I think it would be way simpler to generate the UUID beforehand instead of [relying on `Ecto.Multi`](https://hexdocs.pm/waffle_ecto/filepath-with-id.html). Thoughts?

**Comment by Papipo** [comment]
> For everybody who tried to get this working it is important to use the cast_attachements method within the consume_uploaded_entries. The consume_uploaded_entries will automatically delete the file-path when it is finished so otherwise and {:error, :invalid_file_path} error will be returned.

That collides with this:
https://elixirforum.com/t/forms-with-live-file-input-consume-files-when-saving-even-if-form-is-invalid/56166/2

**Comment by stratigos** [comment]
Thanks to everyone in the history above that helped me find my way with handling the file uploads with today's LiveView of 0.20.15.

Given this: 

> For everybody who tried to get this working it is important to use the cast_attachements method within the consume_uploaded_entries. The consume_uploaded_entries will automatically delete the file-path when it is finished so otherwise and {:error, :invalid_file_path} error will be returned.

The `consume_uploaded_entries` function's closure now can have a return type of `{:postpone, result}`, which allows the tmp file to persist and be handled after `consume_uploaded_entries` finishes (which allows `waffle` to handle the changeset as normal).

https://hexdocs.pm/phoenix_live_view/0.20.14/Phoenix.LiveView.html#consume_uploaded_entries/3

I found success with the following:

```elixir
# I have `max_entries: 1` set, the following expects only one `entry`:
def handle_event(
  "save",
  %{"schema" => schema_params},
  socket
) do
  [schema_params_with_image] =
    consume_uploaded_entries(socket, :image, fn meta, entry ->
      image_plug_upload =
        %Plug.Upload{
          content_type: entry.client_type,
          filename: entry.client_name,
          path: meta.path
        }

      updated_params =
        Map.put(schema_params, "image", image_plug_upload)

      {:postpone, updated_params}
    end)

  save_schema(
    socket,
    socket.assigns.action,
    schema_params_with_image
  )
end   
```
